File: src/main/js/formulas/statistical/COUNT.ts
import { isNotNull } from "../../utils/Other";
import { SheetColumnRowKey } from "../../models/common/SheetColumnRowKey";
import { Complex } from "../../models/common/Types";
import { Converters } from "../../utils/Converters";
import { Counter } from "../../utils/Counter";
import { Mappers } from "../../utils/Mappers";
import { Predicates } from "../../utils/Predicates";
import { Reducers } from "../../utils/Reducers";
import { AbstractFormula } from "../AbstractFormula";
import { FormulaName } from "../FormulaName";

export class COUNT extends AbstractFormula {
  static SELF: COUNT = new COUNT();
  NAME = FormulaName.COUNT;

  internal(origin: SheetColumnRowKey, ...values: Array<Complex>) {
    AbstractFormula.checkAtLeastLength(values.length, 1, this.NAME);
    const counter = new Counter<number>();
    values
      .map(this.lookup)
      .map(Mappers.flattenGridsToArrays) // 1, 2, [null, 1, "0", "Text.", null]
      .map(Mappers.ensureAllAreArrays) // [1], [2], [null, 1, "0", null]
      .reduce(Reducers.join) // [1, 2, null, 1, "0", null]
      .filter(isNotNull) // [1, 2, 1, "0"]
      .filter(Predicates.isNumeric)
      .map(Converters.toNumber)
      .map((value) => counter.count(value)); // [1, 2, 1]
    return counter.getCount();
  }
}

File: src/main/java/io/protobase/f7/formulas/statistical/COUNT.java
package io.protobase.f7.formulas.statistical;

import io.protobase.f7.formulas.AbstractFormula;
import io.protobase.f7.formulas.FormulaName;
import io.protobase.f7.models.GridColumnRowKey;
import io.protobase.f7.utils.Filters;
import io.protobase.f7.utils.Mappers;

import java.util.function.BiFunction;
import java.util.function.Function;
import java.util.stream.Collectors;
import java.util.stream.Stream;

/**
 * The count of number values in a set of data.
 */
public class COUNT extends AbstractFormula {
  public static FormulaName NAME = FormulaName.COUNT;
  public static COUNT SELF = new COUNT();

  public COUNT() {
    super();
  }

  public COUNT(Function<Object, Object> lookup, BiFunction<GridColumnRowKey, Object, Object> collateralLookup) {
    super(lookup, collateralLookup);
  }

  @Override
  public Object internal(GridColumnRowKey origin, Object... values) {
    checkAtLeastLength(values.length, 1, NAME);
    return ((double) Stream.of(values)
        .map(lookup)
        .flatMap(Mappers::toFlatStreamGridFilterOnlyNumberLiterals)
        .filter(Filters::isCoercableToNumeric)
        .collect(Collectors.toList()).size());
  }
}
